### ISSUE TITLE
Network latency is not injected into my application

### ISSUE THREAD

**Issue by iKarthik** [litmuschaos/litmus#1853]
**What happened**:
Executed Pod Network Latency Experiment, i do not see latency injected into the application

**What you expected to happen**:
I have a very simple application deployed tom y K8S cluster. The demo application is based on this [docker image](https://hub.docker.com/r/smorant/kuard-amd64)

**How to reproduce it (as minimally and precisely as possible)**:
Created a basic deployment file that includes the above mentioned image. Created a service account based exactly as specified in this [link](https://docs.litmuschaos.io/docs/pod-network-latency/). Chaos engine manifest also exactly looks like what was in the above mentioned link. I have replaced the app labels wherever appropriate.
When I apply my chaos engine manifest, i can see that k8s creates a pod-network-latency pod as expected. When i look at the logs for the network latency pod, it is able to find the Application pod to inject latency into. Pumba pod appears to get created successfully. I have verified that the network interface of my application pod is eth0. The log files from network latency pod appear to be normal. Here is a sample from it

`W0810 16:13:26.444004       1 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
time="2020-08-10T16:13:26Z" level=info msg="[PreReq]: Getting the ENV for the  experiment"
time="2020-08-10T16:13:26Z" level=info msg="[PreReq]: Updating the chaos result of pod-network-latency experiment (SOT)"
time="2020-08-10T16:13:26Z" level=info msg="The application informations are as follows\n" Namespace=MyNameSPace Label="app=kuard" Ramp Time=0
time="2020-08-10T16:13:26Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (pre-chaos)"
time="2020-08-10T16:13:26Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-08-10T16:13:26Z" level=info msg="The running status of Pods are as follows" Pod=kuard-deployment-6c65d5c8fb-trllp Status=Running
time="2020-08-10T16:13:26Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-08-10T16:13:26Z" level=info msg="The running status of container are as follows" container=kuard Pod=kuard-deployment-6c65d5c8fb-trllp Status=Running
time="2020-08-10T16:13:26Z" level=info msg="[Info]: Details of application under chaos injection" NodeName=ip-10-98-83-22.ec2.internal ContainerName=kuard PodName=kuard-deployment-6c65d5c8fb-trllp
time="2020-08-10T16:13:26Z" level=info msg="[Status]: Checking the status of the helper pod"
time="2020-08-10T16:13:26Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-08-10T16:13:28Z" level=info msg="The running status of Pods are as follows" Status=Running Pod=pumba-netem-judhlx
time="2020-08-10T16:13:28Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-08-10T16:13:28Z" level=info msg="The running status of container are as follows" container=pumba Pod=pumba-netem-judhlx Status=Running
time="2020-08-10T16:13:28Z" level=info msg="[Wait]: Waiting for 300s till the completion of the helper pod"
time="2020-08-10T16:13:28Z" level=info msg="helper pod status: Running"
time="2020-08-10T16:13:29Z" level=info msg="helper pod status: Running"
time="2020-08-10T16:13:30Z" level=info msg="helper pod status: Running"
time="2020-08-10T16:13:31Z" level=info msg="helper pod status: Running"
time="2020-08-10T16:13:32Z" level=info msg="helper pod status: Running"
time="2020-08-10T16:13:33Z" level=info msg="helper pod status: Running"
time="2020-08-10T16:13:34Z" level=info msg="helper pod status: Running"
time="2020-08-10T16:13:35Z" level=info msg="helper pod status: Running"`

I have port-forwarded my application on port 8080. When I access the website localhost:8080, i expect to see network latency, but i don't seem to see any. All XHR requests, that happens as part of the application are able to execute within 50 ms. Is this expected?

**Anything else we need to know?**:

**Comment by ksatchit** [comment]
@iKarthik thank you for trying litmus. Can you share the details of the K8s platform on which this experiment is being run? Also, does pumba logs show any error?

**Comment by iKarthik** [comment]
Hello,

Am on k8s 1.17, deployed on AWS. I do not see any logs from pumba (when i do kubectl logs pumba, i see nothing on stdout)

**Comment by ksatchit** [comment]
cc: @uditgaurav

**Comment by uditgaurav** [comment]
This is interesting. I tried to reproduce this, for that I ran an application pod with the given image (`smorant/kuard-amd64:1`). 
used rbac, experiment and engine from https://hub.litmuschaos.io/generic/pod-network-latency

During the execution of experiment:

```
NAME                               READY   STATUS    RESTARTS   AGE
nginx-6d876b9d9-sqskm              1/1     Running   0          162m
nginx-network-chaos-runner         1/1     Running   0          14s
pod-network-latency-bctptk-8v4px   1/1     Running   0          5s
pumba-netem-ohlxzj                 1/1     Running   0          4s
```

**Pinging the pod ip from the nodes:**

```
[ec2-user@ip-172-31-24-159 ~]$ ping 172.31.20.253
PING 172.31.20.253 (172.31.20.253) 56(84) bytes of data.
64 bytes from 172.31.20.253: icmp_seq=1 ttl=255 time=0.047 ms
From 172.31.24.159 icmp_seq=19 Destination Host Unreachable
From 172.31.24.159 icmp_seq=20 Destination Host Unreachable
From 172.31.24.159 icmp_seq=21 Destination Host Unreachable
From 172.31.24.159 icmp_seq=22 Destination Host Unreachable
From 172.31.24.159 icmp_seq=23 Destination Host Unreachable
From 172.31.24.159 icmp_seq=24 Destination Host Unreachable
From 172.31.24.159 icmp_seq=25 Destination Host Unreachable
From 172.31.24.159 icmp_seq=26 Destination Host Unreachable
From 172.31.24.159 icmp_seq=27 Destination Host Unreachable
From 172.31.24.159 icmp_seq=28 Destination Host Unreachable
From 172.31.24.159 icmp_seq=32 Destination Host Unreachable
From 172.31.24.159 icmp_seq=33 Destination Host Unreachable
From 172.31.24.159 icmp_seq=34 Destination Host Unreachable
From 172.31.24.159 icmp_seq=35 Destination Host Unreachable
From 172.31.24.159 icmp_seq=36 Destination Host Unreachable
From 172.31.24.159 icmp_seq=37 Destination Host Unreachable
From 172.31.24.159 icmp_seq=38 Destination Host Unreachable
From 172.31.24.159 icmp_seq=39 Destination Host Unreachable
From 172.31.24.159 icmp_seq=41 Destination Host Unreachable
From 172.31.24.159 icmp_seq=42 Destination Host Unreachable
From 172.31.24.159 icmp_seq=43 Destination Host Unreachable
From 172.31.24.159 icmp_seq=44 Destination Host Unreachable
From 172.31.24.159 icmp_seq=45 Destination Host Unreachable
From 172.31.24.159 icmp_seq=46 Destination Host Unreachable
From 172.31.24.159 icmp_seq=47 Destination Host Unreachable
From 172.31.24.159 icmp_seq=48 Destination Host Unreachable
From 172.31.24.159 icmp_seq=49 Destination Host Unreachable
From 172.31.24.159 icmp_seq=50 Destination Host Unreachable
From 172.31.24.159 icmp_seq=51 Destination Host Unreachable
From 172.31.24.159 icmp_seq=52 Destination Host Unreachable
From 172.31.24.159 icmp_seq=53 Destination Host Unreachable
From 172.31.24.159 icmp_seq=57 Destination Host Unreachable
From 172.31.24.159 icmp_seq=58 Destination Host Unreachable
64 bytes from 172.31.20.253: icmp_seq=2 ttl=255 time=59995 ms
64 bytes from 172.31.20.253: icmp_seq=3 ttl=255 time=59995 ms
From 172.31.24.159 icmp_seq=59 Destination Host Unreachable
From 172.31.24.159 icmp_seq=60 Destination Host Unreachable
From 172.31.24.159 icmp_seq=61 Destination Host Unreachable
From 172.31.24.159 icmp_seq=62 Destination Host Unreachable
64 bytes from 172.31.20.253: icmp_seq=4 ttl=255 time=59996 ms
64 bytes from 172.31.20.253: icmp_seq=5 ttl=255 time=59997 ms
64 bytes from 172.31.20.253: icmp_seq=6 ttl=255 time=60005 ms
From 172.31.24.159 icmp_seq=63 Destination Host Unreachable
From 172.31.24.159 icmp_seq=64 Destination Host Unreachable
From 172.31.24.159 icmp_seq=65 Destination Host Unreachable
64 bytes from 172.31.20.253: icmp_seq=7 ttl=255 time=60001 ms
64 bytes from 172.31.20.253: icmp_seq=8 ttl=255 time=59993 ms
64 bytes from 172.31.20.253: icmp_seq=9 ttl=255 time=59995 ms
64 bytes from 172.31.20.253: icmp_seq=10 ttl=255 time=60008 ms
From 172.31.24.159 icmp_seq=66 Destination Host Unreachable
From 172.31.24.159 icmp_seq=67 Destination Host Unreachable
From 172.31.24.159 icmp_seq=68 Destination Host Unreachable
From 172.31.24.159 icmp_seq=69 Destination Host Unreachable
64 bytes from 172.31.20.253: icmp_seq=11 ttl=255 time=59997 ms
64 bytes from 172.31.20.253: icmp_seq=12 ttl=255 time=59995 ms
64 bytes from 172.31.20.253: icmp_seq=13 ttl=255 time=60002 ms
64 bytes from 172.31.20.253: icmp_seq=14 ttl=255 time=60002 ms
From 172.31.24.159 icmp_seq=70 Destination Host Unreachable
From 172.31.24.159 icmp_seq=71 Destination Host Unreachable
From 172.31.24.159 icmp_seq=72 Destination Host Unreachable
From 172.31.24.159 icmp_seq=73 Destination Host Unreachable
64 bytes from 172.31.20.253: icmp_seq=15 ttl=255 time=60007 ms
64 bytes from 172.31.20.253: icmp_seq=16 ttl=255 time=60005 ms
64 bytes from 172.31.20.253: icmp_seq=17 ttl=255 time=60005 ms
64 bytes from 172.31.20.253: icmp_seq=18 ttl=255 time=59997 ms
64 bytes from 172.31.20.253: icmp_seq=121 ttl=255 time=0.041 ms
64 bytes from 172.31.20.253: icmp_seq=122 ttl=255 time=0.042 ms
64 bytes from 172.31.20.253: icmp_seq=123 ttl=255 time=0.053 ms
64 bytes from 172.31.20.253: icmp_seq=124 ttl=255 time=0.054 ms
64 bytes from 172.31.20.253: icmp_seq=125 ttl=255 time=0.045 ms
64 bytes from 172.31.20.253: icmp_seq=126 ttl=255 time=0.042 ms
64 bytes from 172.31.20.253: icmp_seq=127 ttl=255 time=0.053 ms
```
So I'm able to see some Network Latency Into Application Pod

cc: @iKarthik @ksatchit

**Comment by iKarthik** [comment]
@uditgaurav Thanks for following up. I replaced the experiment with the latest version from the link you sent above. When i run it, pumba pod is stuck during container creation. It immediately stops with ContainerCannotRun status. Here is what i see in pumba logs

```
Events:
  Type     Reason   Age    From                                  Message
  ----     ------   ----   ----                                  -------
  Normal   Pulling  5m32s  kubelet, ip-10-98-81-69.ec2.internal  Pulling image "gaiaadm/pumba:0.6.5"
  Normal   Pulled   5m32s  kubelet, ip-10-98-81-69.ec2.internal  Successfully pulled image "gaiaadm/pumba:0.6.5"
  Normal   Created  5m32s  kubelet, ip-10-98-81-69.ec2.internal  Created container pumba
  Warning  Failed   5m31s  kubelet, ip-10-98-81-69.ec2.internal  Error: failed to start container "pumba": Error response from daemon: OCI runtime create failed: container_linux.go:349: starting container process caused "exec: \"pumba\": executable file not found in $PATH": unknown
```
Does this mean anything to you? Clearly there is some issue with starting the pumba container, dont know what it is

In terms of workflow, here is what i did

- I cleared all my existing experiments
- Created a single experiment from the link above
- Created rbac for this experiment
- Created chaosengine for this experiment

**Comment by uditgaurav** [comment]
@iKarthik this is due to lib image `gaiaadm/pumba:0.6.5`. In litmus `1.6.2` all lib images and experiment images are merged into one image that is `litmuschaos/go-runner:1.6.2` so that we don't need to pull multiple images for running a single experiment. 

You can try replacing lib image with `litmuschaos/go-runner:1.6.2` or simply use the below chaosengine while using latest version:

```yaml
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name: nginx-network-chaos
  namespace: default
spec: 
  # It can be delete/retain
  jobCleanUpPolicy: 'delete'
  # It can be true/false
  annotationCheck: 'true'
  # It can be active/stop
  engineState: 'active'
  #ex. values: ns1:name=percona,ns2:run=nginx 
  auxiliaryAppInfo: ''
  monitoring: false
  appinfo: 
    appns: 'default'
    # FYI, To see app label, apply kubectl get pods --show-labels
    applabel: 'app=nginx'
    appkind: 'deployment'
  chaosServiceAccount: pod-network-latency-sa
  experiments:
    - name: pod-network-latency
      spec:
        components:
          env:
            #Container name where chaos has to be injected
            - name: TARGET_CONTAINER
              value: 'nginx' 

            #Network interface inside target container
            - name: NETWORK_INTERFACE
              value: 'eth0'     

            - name: LIB_IMAGE
              value: 'litmuschaos/go-runner:1.6.2'

            - name: NETWORK_LATENCY
              value: '60000'

            - name: TOTAL_CHAOS_DURATION
              value: '60' # in seconds
```

Change the ENVs acc. (appinfo and namespace and target container name).

**Comment by iKarthik** [comment]
@uditgaurav With your file above, i was able to get the pods up and running, so thanks for that. But am still seeing that no latency was injected into the pod. This is how am trying to validate it

1. I portforward kuard deployment to port 8080 on my local machine
2. When i access localhost:8080/healthy endpoint on my local machine, i expect to see slow response from the deployment, since latency has been injected into the pod but instead all access to that endpoint happens in under 50 ms.

Please let me know if this method of verification is not going to work if so why. (I don't have a ingress rule defined to be able to access that deployment from outside the cluster at the moment)

**Comment by suhrud-kumar** [comment]
Hi @uditgaurav @ksatchit 

I'm getting below error with latest version of experiment/libraries
```
k describe po pumba-netem-h3it3k -nlitmus-chaos
---
Events:
  Type     Reason   Age   From                                 Message
  ----     ------   ----  ----                                 -------
  Normal   Pulling  88s   kubelet, node-10  Pulling image "litmuschaos/go-runner:1.6.2"
  Normal   Pulled   43s   kubelet, node-10  Successfully pulled image "litmuschaos/go-runner:1.6.2"
  Normal   Created  18s   kubelet, node-10  Created container pumba
  Warning  Failed   15s   kubelet, node-10 Error: failed to start container "pumba": Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"netem\": executable file not found in $PATH": unknown
```

this is my experiment yaml
```
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name:  prom-chaos-network-latency
  namespace: litmus-chaos
spec:
  # It can be delete/retain
  jobCleanUpPolicy: 'delete'
  # It can be true/false
  annotationCheck: 'false'
  # It can be active/stop
  engineState: 'active'
  #ex. values: ns1:name=percona,ns2:run=nginx
  auxiliaryAppInfo: ''
  monitoring: false
  appinfo:
    appns: 'monitoring'
    # FYI, To see app label, apply kubectl get pods --show-labels
    applabel: 'app=prometheus'
    applabel: 'prometheus=prometheus-operator-prometheus'
    appkind: 'statefulset'
  chaosServiceAccount: litmuschaos-admin
  experiments:
    - name: pod-network-latency
      spec:
        components:
          env:
            #Container name where chaos has to be injected
            - name: TARGET_CONTAINER
              value: 'prometheus'

            #Network interface inside target container
            - name: NETWORK_INTERFACE
              value: 'eth0'

            - name: LIB_IMAGE
              value: 'litmuschaos/go-runner:1.6.2'

            - name: NETWORK_LATENCY
              value: '1000'

            - name: TOTAL_CHAOS_DURATION
              value: '120' # in seconds
```

Any idea ?

**Comment by iKarthik** [comment]
> @uditgaurav With your file above, i was able to get the pods up and running, so thanks for that. But am still seeing that no latency was injected into the pod. This is how am trying to validate it
> 
> 1. I portforward kuard deployment to port 8080 on my local machine
> 2. When i access localhost:8080/healthy endpoint on my local machine, i expect to see slow response from the deployment, since latency has been injected into the pod but instead all access to that endpoint happens in under 50 ms.
> 
> Please let me know if this method of verification is not going to work if so why. (I don't have a ingress rule defined to be able to access that deployment from outside the cluster at the moment)

@uditgaurav  I created the necessary ingress rules to access the endpoints outside the cluster. Now am able to see the latency. But oddly enough when i set the NETWORK_LATENCY value to just 400 (milliseconds), i get gateway time out errors (502, 504) when accessing the service when i was just expecting 400 ms latency added to the response duration.

When I delete the network latency experiment, am able to access my service normally (within a few 100 ms). Any idea whats going on here?

**Comment by ksatchit** [comment]
@iKarthik , just for additional data, can you check if this behaviour is observed even when the additional latency (injected value) is of a smaller range (50ms-100ms)? Just wanted to see if there is a percentage of latency added at the gateway itself as we go.

**Comment by ksatchit** [comment]
> Hi @uditgaurav @ksatchit
> 
> I'm getting below error with latest version of experiment/libraries
> 
> ```
> k describe po pumba-netem-h3it3k -nlitmus-chaos
> ---
> Events:
>   Type     Reason   Age   From                                 Message
>   ----     ------   ----  ----                                 -------
>   Normal   Pulling  88s   kubelet, node-10  Pulling image "litmuschaos/go-runner:1.6.2"
>   Normal   Pulled   43s   kubelet, node-10  Successfully pulled image "litmuschaos/go-runner:1.6.2"
>   Normal   Created  18s   kubelet, node-10  Created container pumba
>   Warning  Failed   15s   kubelet, node-10 Error: failed to start container "pumba": Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"netem\": executable file not found in $PATH": unknown
> ```
> 
> this is my experiment yaml
> 
> ```
> apiVersion: litmuschaos.io/v1alpha1
> kind: ChaosEngine
> metadata:
>   name:  prom-chaos-network-latency
>   namespace: litmus-chaos
> spec:
>   # It can be delete/retain
>   jobCleanUpPolicy: 'delete'
>   # It can be true/false
>   annotationCheck: 'false'
>   # It can be active/stop
>   engineState: 'active'
>   #ex. values: ns1:name=percona,ns2:run=nginx
>   auxiliaryAppInfo: ''
>   monitoring: false
>   appinfo:
>     appns: 'monitoring'
>     # FYI, To see app label, apply kubectl get pods --show-labels
>     applabel: 'app=prometheus'
>     applabel: 'prometheus=prometheus-operator-prometheus'
>     appkind: 'statefulset'
>   chaosServiceAccount: litmuschaos-admin
>   experiments:
>     - name: pod-network-latency
>       spec:
>         components:
>           env:
>             #Container name where chaos has to be injected
>             - name: TARGET_CONTAINER
>               value: 'prometheus'
> 
>             #Network interface inside target container
>             - name: NETWORK_INTERFACE
>               value: 'eth0'
> 
>             - name: LIB_IMAGE
>               value: 'litmuschaos/go-runner:1.6.2'
> 
>             - name: NETWORK_LATENCY
>               value: '1000'
> 
>             - name: TOTAL_CHAOS_DURATION
>               value: '120' # in seconds
> ```
> 
> Any idea ?

Figured with @suhrud-kumar from a slack conversation that the go-runner image tag was `latest` in ChaosExperiment CR (CR having been picked off the master branch). This has since been changed to 1.6.2 (now sourced from the versioned hub) & the experiment is now seen to execute as expected.

**Comment by Asgoret** [comment]
Wel...I found another bug ;D
I install litmus via documentation and try network experiment, BUT pumba got wrong regexp information.
Some technical details:
Deployment name: `alertproxy`
Pod name: `alertproxy`
Container name: `alertproxy`
Docker container: `k8s_alertproxy_alertproxy-c8f9ff594-vrfxs_test_7c621948-13ba-11eb-8553-005056011328_0`
Technical pod container: `k8s_POD_alertproxy-c8f9ff594-vrfxs_test_7c621948-13ba-11eb-8553-005056011328_0`
Pumba regexp: `k8s_POD_alertproxy-c8f9ff594-vrfxs_test`
So the main idea is that the pumba got wrong regexp value. It must be `k8s_alertproxy_alertproxy` not `k8s_POD_alertproxy`

Cloud: OKD 3.11

cc @suhrud-kumar @uditgaurav

**Comment by ksatchit** [comment]
@Asgoret this is intentional. We are targeting the pause container associated with the app instead of the app itself - to cover for cases where there are liveness probes (which restart the app containers and modify the ID. The pause container OTOH is not restarted/affected. Consider the pause shares the network namespace with app pod, this is a better choice to help unhindered revert of chaos once the chaos duration is complete)

**Comment by ksatchit** [comment]
> @iKarthik , just for additional data, can you check if this behaviour is observed even when the additional latency (injected value) is of a smaller range (50ms-100ms)? Just wanted to see if there is a variable percentage of latency added at the gateway/ingress controller itself as we go.

I suppose the original issue of not being able to inject latency was resolved for @iKarthik . The question seems to be an unexpected amount of latency. Can you confirm this: https://github.com/litmuschaos/litmus/issues/1853#issuecomment-673918504

Also,  recent versions of the experiment also support: (a) litmus lib for n/w latency (you don't have to pull pumba images) (b) restrict traffic to specific destination IPs/hosts

**Comment by ksatchit** [comment]
Closing this issue due to lack of recent updates. Please re-open if things continue to not work with latest litmus releases.
